server/src/main/java/org/opensearch/search/resource_limit_group/tracker/ResourceLimitsGroupResourceUsageTrackerService.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.resource_limit_group.tracker;
+
+import org.opensearch.cluster.metadata.ResourceLimitGroup;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.core.tasks.resourcetracker.TaskResourceUsage;
+import org.opensearch.search.resource_limit_group.ResourceLimitGroupPruner;
+import org.opensearch.search.resource_limit_group.ResourceLimitGroupTask;
+import org.opensearch.search.resource_limit_group.cancellation.CancellableTaskSelector;
+import org.opensearch.search.resource_limit_group.cancellation.ResourceLimitGroupTaskCanceller;
+import org.opensearch.tasks.Task;
+import org.opensearch.tasks.TaskCancellation;
+import org.opensearch.tasks.TaskManager;
+import org.opensearch.tasks.TaskResourceTrackingService;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+
+/**
+ * This class tracks requests per resourceLimitGroups
+ */
+public class ResourceLimitsGroupResourceUsageTrackerService extends ResourceLimitGroupTaskCanceller
+    implements
+        TaskManager.TaskEventListeners,
+        ResourceLimitGroupResourceUsageTracker,
+        ResourceLimitGroupPruner {
+
+    private static final String CPU = "cpu";
+    private static final String JVM = "jvm";
+    private static final List<String> TRACKED_RESOURCES = List.of(JVM);
+    private static final int numberOfAvailableProcessors = Runtime.getRuntime().availableProcessors();
+    private static final long totalAvailableJvmMemory = Runtime.getRuntime().totalMemory();
+    private final LongSupplier timeNanosSupplier;
+    /**
+     * ResourceLimitGroup ids which are marked for deletion in between the
+     * {@link org.opensearch.search.resource_limit_group.ResourceLimitGroupService} runs
+     */
+    private List<String> toDeleteResourceLimitGroups;
+    private List<ResourceLimitGroup> activeResourceLimitGroups;
+    /**
+     * This var will hold the sandbox level resource usage as per last run of
+     * {@link org.opensearch.search.resource_limit_group.ResourceLimitGroupService}
+     */
+    private Map<String, Map<String, Double>> resourceUsage;
+    /**
+     * We are keeping this as instance member as we will need this to identify the contended resource limit groups
+     * resourceLimitGroupName -> List&lt;ResourceLimitGroupTask&gt;
+     */
+    private Map<String, List<Task>> resourceLimitGroupTasks;
```

Comment:
Should this be hardcoded to Cancellable task ? would we ever tag a non Cancellable task type to a ResourceLimitGroup ?

Reply:
No